jni-gen/src/utils.rs

```diff
@@ -82,3 +82,44 @@ pub fn java_identifier_to_rust(name: &str) -> String {
     // identifier can only be composed of [a-zA-Z&_] characters - https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8
     name.replace('_', "__").replace('$', "_dollar_")
 }
+
+pub fn is_signature_of_class_type(signature: &str) -> bool {
+    signature.starts_with('L')
+}
+
+/// Generates a runtime check that the object is of the expected class - applies only to object types (starting with L)
+/// If this value is also the returned value in the setter, unwrap it first (since it is of type Result<T>)
+fn generate_type_check(variable_name: &str, variable_type_name: &str, is_result: bool) -> String {
+    let mut type_check: Vec<String> = vec![];
+    type_check.push("    debug_assert!(".to_string());
+    type_check.push(format!(
+        "        self.env.is_instance_of({variable_name}, self.env.find_class({variable_type_name})?)?"
+    ));
+    type_check.push("    );".to_string());
+
+    if is_result {
+        let indented_type_check = type_check
+            .iter()
+            .map(|x| format!("    {x}"))
+            .collect::<Vec<_>>()
+            .join("\n");
+
+        vec![
+            "    if let Ok(result) = result {".to_string(),
+            indented_type_check,
+            "    }".to_string(),
+        ]
+    } else {
+        type_check
+    }
+    .join("\n")
+        + "\n"
+}
+
+pub fn generate_result_type_check(variable_type_name: &str) -> String {
```

Comment:
I think the interface/names/docs changes I made are unclear, sorry for the confusion.

I decided to declare all the class names at the beginning of the functions (e.g. arg_arg0_class_name)
![image](https://user-images.githubusercontent.com/8255346/223518742-1f064ca2-ae30-47e8-94e0-1d35c4b8494b.png)
or 
![image](https://user-images.githubusercontent.com/8255346/223518848-0df6e06c-aeb2-40a4-9327-ad754b5d6d28.png)
If the parameter/return type is not of a class type then both generating the variable with a class name for the respective parameter/return types and generating the check are skipped (in the second one parameter is skipped)

Then in the check, the name of the variable is used, rather than the string literal:
![image](https://user-images.githubusercontent.com/8255346/223518659-94a8a11d-3429-4e37-96e8-04beec7d2fa3.png)

 
Before both ways were used (e.g. sometimes class_name, sometimes "java/lang/String") and I was considering hardcoding the string literals with the class names there but decided to rather use variables for clarity (once the class names are put there as literals it takes more time to backtrack how did they get there and if they are there by an accident). I also thought to allow to use both (by passing an enum) but thought it might be overengineering.

This has the problem that at the time of the generation I the `generate_type_check` function cannot do the check.

BTW there is also a bug (get check the return type with signature, not the class name).

Reply:
Thanks for the explanation. I didn't notice that the argument of `generate_type_check` is now a `/`-separated class name instead of a signature. Can you write this in the doc comment? I agree that an enum would be an overengineering.

Reply:
> This has the problem that at the time of generation the `generate_type_check` function cannot do the check.

Do you mean the `debug_assert!(..)` that I suggested to check that the signature is of a Java class? Sure, if the argument is a class name then that check does not make sense.

Reply:
Changed the parameter names